### PR TITLE
Initial target_fragsfilter conversion support

### PR DESCRIPTION
**Pull request by Mac-tf2**

### CODE REVIEW

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 1]
```diff
@@ -129,9 +137,6 @@ public void Convert()
 					case "trigger_teleport":
 						ConvertTriggerTeleport(entity);
 						break;
-					case "misc_teleporter_dest":
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Why was this removed?

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 2]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+					delay = delay,
+					fireOnce = -1
+				};
+				entity.connections.Add(connection);
+			}
+		}
+
+		private void ConvertTargetScore(Entity entity, Entity targetScore, string output, float delay)
+		{
+			if (sourceEntities.FirstOrDefault(x => x.ClassName == "math_counter") == null) // Check if math_counter exists
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Use `!sourceEntities.Any(x => x.ClassName == "math_counter")` instead

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 3]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+				CreateMathCounter();
+
+			if (!float.TryParse(targetScore["count"], out var count))
+				count = 1;
+
+			ModifyMathCounter(entity, output, "Add", $"{count}", delay);
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Replace `$"{count}"` with `count.ToString()`

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 4]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+		}
+
+		private Entity CreateLogicCase()
+		{
+			var logicCase = new Entity();
+			logicCase.ClassName = "logic_case";
+			logicCase.Name = "logiccase";
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: You should replace this with a constant that has a more unique name:
`private const string MOMENTUM_LOGIC_CASE = "_momentum_logic_case_";`

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 5]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+
+			for (var i = 1; i <= 16; i++) // Logic_case supports 16 different outputs
+			{
+				var caseNum = $"case{i:D2}";
+				logicCase[$"{caseNum}"] = $"{i}";
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: I think the string interpolation is unnecessary here.
`logicCase[caseNum] = i;`. If that doesn't work, then just add `.ToString()`

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 6]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+			counter.ClassName = "math_counter";
+			counter.Name = "mathcounter";
+			counter["startvalue"] = "0";
+			counter["min"] = "0";
+			counter["max"] = "16";
+
+			var connection = new Entity.EntityConnection()
+			{
+				name = "OutValue",
+				target = "logiccase",
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Replace this with the constant value that I mentioned on line 679.

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 7]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+			};
+			entity.connections.Add(connection);
+
+			var match = false;
+			var spawnflags = (TargetFragsFilterFlags)targetFragsFilter.Spawnflags;
+
+			if (spawnflags.HasFlag(TargetFragsFilterFlags.Reset)) // Reset frags to 0
+				ModifyMathCounter(targetFragsFilter, "OnTrigger", "SetValue", "0", delay);
+			else if (spawnflags.HasFlag(TargetFragsFilterFlags.Remover)) // Remove frags when used
+				ModifyMathCounter(targetFragsFilter, "OnTrigger", "Subtract", $"{frags}", delay);
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Replace `$"{frags}"` with `frags.ToString()`

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 8]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+				match = true;
+
+			AddLogicCaseOutput(targetFragsFilter.Name, frags, match);
+
+			ConvertTriggerTargetsRecursive(targetFragsFilter, targetFragsFilter, 0, new HashSet<Entity>());
+		}
+
+		private void AddLogicCaseOutput(string targetName, int frags, bool match)
+		{
+			var logicCase = sourceEntities.FirstOrDefault(x => x.ClassName == "logic_case") ?? CreateLogicCase();
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: I think `sourceEntities.Find` is slightly more performant and does the same thing.

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 9]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+
+			var min = frags;
+			var max = match ? frags : 16; // Either force frags to match case number on true, else allow any cases over the frag count to trigger
+
+			for (var i = min; i <= max; i++)
+			{
+				var caseNum = $"case{i:D2}";
+
+				logicCase[$"{caseNum}"] = $"{i}";
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Remove the string interpolation here since you're not concatenating strings with other data types.

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 10]
```diff
@@ -597,23 +612,192 @@ private void ConvertTriggerTargetsRecursive(Entity trigger, Entity entity, float
...
+			sourceEntities.Add(logicCase);
+
+			return logicCase;
+		}
+
+		private void CreateMathCounter()
+		{
+			var counter = new Entity();
+			counter.ClassName = "math_counter";
+			counter.Name = "mathcounter";
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: This should be a constant with a more unique name.

**Comment by Panzerhandschuh** [comment]
Wouldn't our collectibles system replace a lot of this logic?

**Comment by Mac-tf2** [comment]
I looked into collectibles but you can't differentiate between matching values and passing the minimum required value. It's also not possible to filter only some of the triggered inputs, it's either all or none. Defrag lets you have entities that fire before hitting the fragsfilter check. ``trigger_multiple > target_speed > target_fragsfilter > target_give`` would always fire the target_speed, but not the target_give unless you pass the frags check.

**Review by Panzerhandschuh — APPROVED** [review]
